src/optimizers.jl

```diff
@@ -1,239 +1,147 @@
-abstract type Optimizer end
+PastaQ.state(optimizer, model::LPDO) = 
+  optimizerstate(optimizer, getparameters(model))
 
-struct SGD <: Optimizer
-  η::Float64
-  γ::Float64
-  v::Vector{ITensor}
-end
+PastaQ.state(optimizer, model::MPS) = 
+  PastaQ.state(optimizer, LPDO(model))
 
-""" 
-    SGD(L::LPDO;η::Float64=0.01,γ::Float64=0.0)
+PastaQ.state(optimizer, model::MPO) = 
+  PastaQ.state(optimizer, LPDO(model))
 
-Stochastic gradient descent with momentum.
 
-# Parameters
-  - `η`: learning rate
-  - `γ`: friction coefficient
-  - `v`: "velocity"
 """
-function SGD(; η::Float64=0.01, γ::Float64=0.0)
-  v = ITensor[]
-  return SGD(η, γ, v)
-end
-
-Base.copy(opt::SGD) = SGD(opt.η, opt.γ, copy(opt.v))
+    update!(model, grads, optimizer)
 
+Update a tensor network model
 """
-    update!(L::LPDO,∇::Array,opt::SGD; kwargs...)
-
-Update parameters with SGD.
+function PastaQ.update!(model, grads, optimizer)
+  L = deepcopy(model)
+  opt, st = optimizer
+  θ, ∇ = getparameters(L; ∇ = copy(grads))
+  st, θ′ = Optimisers.update(opt, st, θ, ∇)
+  
+  setparameters!(L, θ′)
+  model.X[:] = L.X
+  return model
+end
 
-1. `vⱼ = γ * vⱼ - η * ∇ⱼ`: integrated velocity
-2. `θⱼ = θⱼ + vⱼ`: parameter update
-"""
-function update!(L::LPDO, ∇::Array, opt::SGD; kwargs...)
-  M = L.X
-  if isempty(opt.v)
-    for j in 1:length(M)
-      push!(opt.v, ITensor(zeros(size(M[j])), inds(M[j])))
+function gettensorindices(M::MPS, site::Int)
```

Comment:
Could you explain how this function is used? I know it has to do with converting back and forth from the vector format that is needed for `Optimisers`.

Reply:
This function returns the indices of a particular MPS/MPO tensor at site `site`. The order of these indices is then used to parse the tensor into a 1d vector. We need to standardization because the order of the indices in the tensor network to be optimized is not necessarily the same as the gradient being generated. We don't want to rely on the index ordering to be the same, but at some point this has to happen.